src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1973,29 +1988,140 @@ internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExp
+        internal ImmutableArray<IInterpolatedStringContentOperation> CreateBoundInterpolatedStringContentOperation(ImmutableArray<BoundExpression> parts, InterpolatedStringHandlerData? data)
         {
-            var builder = ArrayBuilder<IInterpolatedStringContentOperation>.GetInstance(parts.Length);
-            foreach (var part in parts)
+            return data is { PositionInfo: var positionInfo } ? createHandlerInterpolatedStringContent(positionInfo) : createNonHandlerInterpolatedStringContent();
+
+            ImmutableArray<IInterpolatedStringContentOperation> createNonHandlerInterpolatedStringContent()
             {
-                if (part.Kind == BoundKind.StringInsert)
+                var builder = ArrayBuilder<IInterpolatedStringContentOperation>.GetInstance(parts.Length);
+                foreach (var part in parts)
                 {
-                    builder.Add((IInterpolatedStringContentOperation)Create(part));
+                    if (part.Kind == BoundKind.StringInsert)
+                    {
+                        builder.Add((IInterpolatedStringContentOperation)Create(part));
+                    }
+                    else
+                    {
+                        builder.Add(CreateBoundInterpolatedStringTextOperation((BoundLiteral)part));
+                    }
                 }
-                else
+
+                return builder.ToImmutableAndFree();
+            }
+
+            ImmutableArray<IInterpolatedStringContentOperation> createHandlerInterpolatedStringContent(ImmutableArray<(bool IsLiteral, bool HasAlignment, bool HasFormat)> positionInfo)
+            {
+                // For interpolated string handlers, we want to deconstruct the `AppendLiteral`/`AppendFormatted` calls into
+                // their relevant components.
+                // https://github.com/dotnet/roslyn/issues/54505 we need to handle interpolated strings used as handler conversions.
+
+                Debug.Assert(parts.Length == positionInfo.Length);
+                var builder = ArrayBuilder<IInterpolatedStringContentOperation>.GetInstance(parts.Length);
+
+                for (int i = 0; i < parts.Length; i++)
                 {
-                    builder.Add(CreateBoundInterpolatedStringTextOperation((BoundLiteral)part));
+                    var part = parts[i];
+                    var currentPosition = positionInfo[i];
+
+                    BoundExpression value;
+                    BoundExpression? alignment;
+                    BoundExpression? format;
+
+                    switch (part)
+                    {
+                        case BoundCall call:
+                            (value, alignment, format) = getCallInfo(call.Arguments, call.ArgumentNamesOpt, currentPosition);
+                            break;
+
+                        case BoundDynamicInvocation dynamicInvocation:
```

Comment:
I couldn't identify a test that hits this case.

Reply:
We do not. It's on my plan did Tuesday.

Reply:
And I did not remove the prototype comment related to dynamic because of this.

Reply:
> It's on my plan for Tuesday.

Sounds good. Thanks
My understanding is that we don't support dynamic scenario in binding or lowering, so I was surprised to see some support in IOp.

Reply:
We don't support emitting code, but that doesn't mean the IOp tree creation should fail.